File: Python/HeapSort.py
'''
Explanation:
1. The `heapify` function takes three parameters: the array `arr`, the size of the heap `n`, and the index of the current node `i`.
2. It first assumes the largest element is at the current index `i`.
3. It calculates the indices of the left and right children of the current node.
4. It checks if the left child exists and is greater than the current largest element.
5. If so, it updates the index of the largest element.
6. It then checks if the right child exists and is greater than the current largest element.
7. If so, it updates the index of the largest element.
8. If the largest element is not the current node, it swaps the values and recursively calls `heapify` on the affected subtree.
9. The `heap_sort` function takes an array `arr` and performs the Heap Sort algorithm.
10. It initializes the size of the heap `n` to the length of the array.
11. It builds a max-heap from the array by calling `heapify` on each non-leaf node in reverse order.
12. It extracts elements from the heap one by one by swapping the root (maximum element) with the last unsorted element.
13. After each swap, it calls `heapify` on the reduced heap size to ensure the max-heap property is maintained.
14. Finally, it returns the sorted array.
15. In the sample usage, an input array `[12, 11, 13, 5, 6, 7]` is sorted using heap sort.
16. The sorted array `[5, 6, 7, 11, 12, 13]` is printed.

Time Complexity:
The time complexity of heap sort is O(n log n) in all cases, where n is the number of elements in the array. Both the `heapify` function and the repeated swapping take O(log n) time, and they are performed n times.

Space Complexity:
The space complexity of heap sort is O(1) since the sorting is performed in-place, without using any additional data structures that depend on the input size.

Sample Input:
input_array = [12, 11, 13, 5, 6, 7]

Sample Output:
Sorted array: [5, 6, 7, 11, 12, 13]

In the above example, the input array `[12, 11, 13, 5, 6, 7]` is sorted using heap sort, and the resulting sorted array is `[5, 6, 7, 11, 12, 13]`.
'''

def heapify(arr, n, i):
    largest = i  # Initialize the largest element as the root
    left = 2 * i + 1  # Left child
    right = 2 * i + 2  # Right child

    # Check if the left child of the root exists and is greater than the root
    if left < n and arr[i] < arr[left]:
        largest = left

    # Check if the right child of the root exists and is greater than the root and left child
    if right < n and arr[largest] < arr[right]:
        largest = right

    # Swap the root with the largest element, if necessary
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  # Swap
        heapify(arr, n, largest)

def heap_sort(arr):
    n = len(arr)

    # Build a max-heap from the array
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)

    # Extract elements one by one from the heap
    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]  # Swap
        heapify(arr, i, 0)

    return arr

# Sample usage
input_array = [12, 11, 13, 5, 6, 7]
sorted_array = heap_sort(input_array)
print("Sorted array:", sorted_array)

File: Java/Sorting/HeapSort.java
/*
Explanation of the approach:
1. The heap sort algorithm uses a binary heap data structure to sort the elements.
2. First, we build a max heap from the input array.
3. To build the max heap, we start from the parent of the last element and heapify all the elements up to the root of the heap.
4. The heapify operation ensures that the largest element is at the root of the heap.
5. Once the max heap is built, we extract elements one by one from the heap and move them to the end of the array.
6. After each extraction, we heapify the remaining heap to maintain the max heap property.
7. Finally, the array is sorted in ascending order.

Time Complexity: 
- Building the max heap takes O(n) time.
- Extraction of each element and heapify operation takes O(log n) time.
- Therefore, the overall time complexity of heap sort is O(n log n).

Space Complexity:
- The space complexity of heap sort is O(1) since it does not require any additional space.

Sample Input:
Input: [12, 11, 13, 5, 6, 7]

Sample Output:
Output: [5, 6, 7, 11, 12, 13]
*/

import java.util.Arrays;

public class HeapSort {
    
    public static void heapSort(int[] array) {
        int n = array.length;
        
        // Build max heap
        for (int i = n / 2 - 1; i >= 0; i--) {
            heapify(array, n, i);
        }
        
        // Extract elements from heap
        for (int i = n - 1; i >= 0; i--) {
            // Move current root to end
            int temp = array[0];
            array[0] = array[i];
            array[i] = temp;
            
            // Max heapify the reduced heap
            heapify(array, i, 0);
        }
    }
    
    public static void heapify(int[] array, int n, int i) {
        int largest = i; // Initialize largest as root
        int left = 2 * i + 1;
        int right = 2 * i + 2;
        
        // If left child is larger than root
        if (left < n && array[left] > array[largest]) {
            largest = left;
        }
        
        // If right child is larger than root
        if (right < n && array[right] > array[largest]) {
            largest = right;
        }
        
        // If largest is not root
        if (largest != i) {
            int swap = array[i];
            array[i] = array[largest];
            array[largest] = swap;
            
            // Recursively heapify the affected sub-tree
            heapify(array, n, largest);
        }
    }
    
    public static void main(String[] args) {
        int[] array = {12, 11, 13, 5, 6, 7};
        System.out.println("Input: " + Arrays.toString(array));
        
        heapSort(array);
        
        System.out.println("Output: " + Arrays.toString(array));
    }
}


